### PR TITLE
Close info bars when associated project(s) unload

### PR DESCRIPTION
**Pull request by drewnoakes**
Previously, info bars would remain in VS after solution close, or after project unload. This can lead to confusion when opening a new solution and seeing an error message from the previously open solution.

This change keeps track of which project(s) are associated with each info bar, and monitors the unload event of those projects. Now, when a project unloads, it will remove the info bar if no other remaining projects are also associated with that info bar.

The code was refactored a fair bit, reducing the number of source files. Quite a few refactorings are in their own commits to make review easier. The main change is in 66270deff6939a931aead11bf31f31dd593f9f04.

clear-info-bar-on-solution-change

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/project-system/pull/9608)

### CODE REVIEW

**Review by haileymck — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                {
+                    entry.OnProjectClosed(_project);
+                }
+            }
+        });
+    }
+
+    public async Task ShowInfoBarAsync(string message, ImageMoniker image, CancellationToken cancellationToken, params ImmutableArray<InfoBarAction> actions)
+    {
+        if (await _vsShellServices.IsCommandLineModeAsync(cancellationToken))
```
Comment (LittleLittleCloud) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs:60]: What's command line mode in VS

**Review by LittleLittleCloud — APPROVED** [review]

File: spelling.dic [hunk 2]
```diff
@@ -1,6 +1,7 @@
 appsettings
 args
 async
+asynch
```
Comment (LittleLittleCloud) [inline spelling.dic:4]: Is this change necessary.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Notifications/INonModalNotificationService.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.VisualStudio.Notifications;
 /// <summary>
 ///     Provides methods for displaying non-blocking notifications to the user.
 /// </summary>
-[ProjectSystemContract(ProjectSystemContractScope.ProjectService, ProjectSystemContractProvider.Private, Cardinality = ImportCardinality.ExactlyOne)]
+[ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.Private, Cardinality = ImportCardinality.ExactlyOne)]
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Notifications/INonModalNotificationService.cs:10]: There are several MEF scopes. Global, ProjectServices (similar to global), UnconfiguredProject and ConfiguredProject. There's only one instance of Global and ProjectServices, and there are potentially many instances of UnconfiguredProject and ConfiguredProject.

A component that lives inside UnconfiguredProject is created for each project.

So this change means that the component is now created for every project. Note that this attribute doesn't _control_ the scopes, it just documents it. What actually changes the scope is the `ImportingConstructor` argument list. This PR adds an import of `UnconfiguredProject`, which forces the component to move from an outer scope to the inner scope.

Now that the component has access to a project, we can register the infobars against those projects.

That's a brief discussion of scopes. There's more to say on the topic. We have some docs for scopes in https://github.com/microsoft/VSProjectSystem/blob/master/doc/overview/scopes.md but I don't think they are as helpful as they could be. I hope to improve them some time.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs:60]: If you run `devenv /build MySolution.sln` then VS runs without any UI and builds the solution. We call this "command line mode" and there's no UI. We track this in several places to both:

- Improve performance (avoiding work when there's no UI)
- Prevent hangs (due to waiting for things that will never be available in this mode)

File: spelling.dic (hunk 2 above)
Comment (drewnoakes) [reply spelling.dic:4]: It's not in this PR but it is in a class I was looking at. I generally add words when I find them, to avoiding having warnings in the editor in VS.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    private readonly UnconfiguredProject _project;
+    private readonly IProjectThreadingService _threadingService;
+    private readonly IVsShellServices _vsShellServices;
+    private readonly IVsUIService<SVsShell, IVsShell> _vsShell;
+    private readonly IVsUIService<SVsInfoBarUIFactory, IVsInfoBarUIFactory> _vsInfoBarFactory;
+
+    [ImportingConstructor]
+    public VsInfoBarService(
+        UnconfiguredProject project,
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs:31]: @LittleLittleCloud this is the actual change that moves the scope of the MEF component into UnconfiguredProject scope. The previous service only imported from the Glboal and ProjectServices scopes. Now we import UnconfiguredProject, so we're moved into that scope.
